pandas/core/index.py

```diff
@@ -391,6 +395,11 @@ def _shallow_copy(self, values=None, infer=False, **kwargs):
 
         if infer:
             attributes['copy'] = False
+            if self._infer_as_myclass:
```

Comment:
Below is the behavior of current master. 

```
pd.Index([1., 2., 3.])._shallow_copy([1, 2, 3], infer=True)
# Int64Index([1, 2, 3], dtype='int64')

pd.Index([1., 2., 3.])._shallow_copy([1, 2, 3])
# Float64Index([1, 2, 3], dtype='int64')
```

```
idx = pd.date_range('2011-01-01', freq='D', periods=3)
idx._shallow_copy(idx.asi8, infer=True)
# Int64Index([1293840000000000000, 1293926400000000000, 1294012800000000000], dtype='int64')

idx._shallow_copy(idx.asi8)
# DatetimeIndex(['2011-01-01', '2011-01-02', '2011-01-03'], dtype='datetime64[ns]', freq='D')
```

``.loc`` must ``infer`` because user may pass different dtype.

```
pd.Index([1., 2., 3.])._shallow_copy(['a', 'b', 'c'])
# Float64Index([u'a', u'b', u'c'], dtype='|S1')
# -> NG, incorrect dtype
```

Thus, I understand ``_shallow_copy`` should cover followings. 

- Return the same ``Index`` if values are guaranteed to be the same dtype as the original (without inference)
- Return the correct type of ``Index`` if values can have different dtypes from the original (with inference) 
  - Inference prioritizing original class (used in datetime-like ``Index`` accepting internal repr like ``asi8``)
  - Normal inference (simply pass values to ``Index``)

Class attribute ``_infer_as_myclass`` intends to split the logic as below (or needs to override ``_shallow_copy`` in ``tseries/base``)

```
pd.Index([1., 2., 3.])._shallow_copy(['2011-01', '2011-02', '2011-02'], infer=True)
# Index([u'2011-01', u'2011-02', u'2011-02'], dtype='object')

idx._shallow_copy(['2011-01', '2011-02', '2011-02'], infer=True)
# Index([u'2011-01', u'2011-02', u'2011-02'], dtype='object')
# -> want DatetimeIndex in this case. 
```